src/server/main_service.cc

```diff
@@ -2335,26 +2335,23 @@ void Service::OnClose(facade::ConnectionContext* cntx) {
   ConnectionState& conn_state = server_cntx->conn_state;
 
   if (conn_state.subscribe_info) {  // Clean-ups related to PUBSUB
-    optional<BlockingCounter> token;
-
     if (!conn_state.subscribe_info->channels.empty()) {
-      token = conn_state.subscribe_info->borrow_token;
+      auto token = conn_state.subscribe_info->borrow_token;
       server_cntx->UnsubscribeAll(false);
+
+      // Check that all borrowers finished processing.
+      // token is increased in channel_slice (the publisher side).
+      token.Wait();
     }
 
     if (conn_state.subscribe_info) {
       DCHECK(!conn_state.subscribe_info->patterns.empty());
-      // DCHECK(!token || *token == conn_state.subscribe_info->borrow_token);
 
-      token = conn_state.subscribe_info->borrow_token;
+      auto token = conn_state.subscribe_info->borrow_token;
       server_cntx->PUnsubscribeAll(false);
+      token.Wait();  // Same as above
     }
 
-    // Check that all borrowers finished processing.
-    // The tokens are increased in channel_slice (the publisher side).
-    if (token)
-      token->Wait();
-
```

Comment:
the token must be the same because we can't re-subscribe in-between those two if's as conn_closing = true and a subscribe simply won't be executed, however let's not make the change in this PR... Either we'll introduce refcounts or make blocking counter comparable